src/utils/config.py

```diff
@@ -0,0 +1,238 @@
+                        raise ValueError(f"Overlapping keys in configs: {key}:{subkey}")
+            else:
+                raise ValueError(f"Overlapping keys in configs: {key}")
+
+        else:
+            config2[key] = config1[key]
+    return deepcopy(config2)
+
+
+def _check_has_schema(config_item: dict) -> bool:
+    """Check if a validation config item is a schema.
+
+    This is done by seeing whether the item has the required keys.
+
+    Args:
+        config_item (dict): An item from the validation schema.
+
+    Returns:
+        bool: Whether config item has required keys.
+    """
+    expected_keys = [
+        "singular",
+        "dtype",
+        "accept_nonetype",
+    ]
+    if set(expected_keys).issubset(set(list(config_item.keys()))):
+        return True
+    return False
+
+
+def _validate_path(path: str, schema_config: dict):
+    """Check a path has the extension specified in the config schema.
+
+    Args:
+        path (str): The path to validate.
+        schema_config (dict): The config schema containing the file extension.
+
+    Raises:
+        KeyError: If the validation schema config does not contain the 'filetype' key.
+
+    Returns:
+        None
+    """
+    try:
+        file_ext = schema_config["filetype"]
+    except KeyError:
+        file_ext = None
+    if file_ext:
+        validate_file_extension(path, file_ext)
+
+
+def _validate_numeric(value: Union[float, int], param_nm: str, config: dict):
+    """Validate a numerical value to ensure it falls within a specified range.
+
+    Args:
+        value (Union[float, int]): The numerical value to be validated.
+        param_nm (str): The name of the parameter being validated (for error messages).
+        config (dict): The configuration schema containing the min and max values.
+
+    Raises:
+        ValueError: If the value is greater than the maximum specified in the config.
+        ValueError: If the value is less than the minimum specified in the config.
+    """
+    try:
+        max_value = config["max"]
+    except KeyError:
+        max_value = None
+    try:
+        min_value = config["min"]
+    except KeyError:
+        min_value = None
+    if max_value:
+        if value > max_value:
+            raise ValueError(
+                f"Config value for {param_nm} ({value}) greater than max ({max_value})."
+            )
+    if min_value:
+        if value < min_value:
+            raise ValueError(
+                f"Config value for {param_nm} ({value}) less than min ({min_value})."
+            )
+
+
+def _nulltype_conversion(value: str) -> Union[str, None]:
+    """Account for user disparities in the form of NoneType they pass.
+
+    Args:
+        value (str): The value passed by a user.
+
+    Returns:
+        Union[str, None]: The altered value.
+    """
+    type_defence(value, "value", str)
+    if value.lower() in ["", "null", "none"]:
+        return None
+    return value
+
+
+def _validate_config_items(  # noqa C901
```

Comment:
I needed to put the # noqa here as it said the function is too complex.
As (hopefully) a final task, can you break it up please? It will help readability as well as passing the precommits.

Reply:
@AnneONS This isn't necessarily too complex in terms of branches. Splitting this up would either create a lot of small, one time use functions, or one larger one time use function that would essentially just wrap most of this function. To me, this does not feel worth it.

Reply:
To add to this, I've managed to reduce the complexity within the function, so it should be okay now.

Reply:
Never mind, it still hits at a complexity of 10. I feel this is too low, and should at minimum accept 10. THe function doesnt appear too complex to me as a developer.